### PR TITLE
Maintenance: Implement the ability to abort API fetch requests

### PR DESCRIPTION
**Pull request by GP-Dan-Tovbein**
Ref: https://jira.greenpeace.org/browse/PLANET-6921

## Description
We're currently getting this log error when any Cover, Article or Gallery block is unmounted without finishing the request.

We solve this by implementing a clean up function (i.e. [here](https://github.com/greenpeace/planet4-plugin-gutenberg-blocks/pull/940#discussion_r978820100)) and aborting all the API fetch that were requested by the Block.  

Doing this we will avoid to run a ton of unexpected requests.

```
Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
```

## Testing
1. Create a new page
2. Close the Search pattern modal after 10 seconds, this is depending on your connection, or wait to start loading the block you want to test. For instance, if you want to test the `Articles block`, then you should wait the `Deep Dive Topic` which is using it.
3. Open the console and see the logged errors. You ought to see any related to `cancel all subscriptions and asynchronous tasks`.

### Covers Block (fix and solution)

https://user-images.githubusercontent.com/77975803/192011383-8971f9c9-b818-4eaf-8dde-8c4a6c343e52.mov

https://user-images.githubusercontent.com/77975803/192011276-246409d3-34ad-4d7c-b79f-6266af3fb37f.mov

### Articles block (fix and solution)

https://user-images.githubusercontent.com/77975803/192013772-fd8a6db6-2671-481c-8819-3c808c9a37df.mov

### Gallery Block (fix and solution)

https://user-images.githubusercontent.com/77975803/192029312-b14e9676-2d8f-4baf-8cf2-c5a8830d917d.mov

### CODE REVIEW

File: assets/src/blocks/Covers/useCovers.js [hunk 1]
```diff
@@ -53,26 +54,38 @@ export const useCovers = ({ post_types, tags, cover_type, initialRowsLimit, post
...
+      setController(typeof AbortController === 'undefined' ? undefined : new AbortController());
     }
   }, [cover_type, post_types, tags, posts, layout]);
 
+  useEffect(() => {
+    if(controller) {
+      loadCovers();
+    }
+
+    return () => {
```
Comment (GP-Dan-Tovbein) [inline assets/src/blocks/Covers/useCovers.js:80]: This clean up function is invoked every time the component is being unmounted. You can take a look at the [React Documentation](https://reactjs.org/docs/hooks-effect.html#effects-with-cleanup) for more information.

**Comment by planet-4** [comment]
### Test instance is ready :rocket:

:new_moon: [pandora](https://www-dev.greenpeace.org/test-pandora) | [admin](https://www-dev.greenpeace.org/test-pandora/wp-admin/) | [blocks report](https://www-dev.greenpeace.org/test-pandora/wp-admin/?page=plugin_blocks_report) | [CircleCI](https://app.circleci.com/pipelines/github/greenpeace/planet4-test-pandora) | [composer-local.json](https://github.com/greenpeace/planet4-test-pandora/blob/main/composer-local.json)

:watch: 2022.09.23 16:13:28

File: assets/src/blocks/Articles/useArticlesFetch.js [hunk 2]
```diff
@@ -59,9 +60,23 @@ export const useArticlesFetch = (attributes, postType, postId, baseUrl = null, p
 
   useEffect(() => {
     setDisplayedPosts([]);
-    loadPage(true);
+    setController(typeof AbortController === 'undefined' ? undefined : new AbortController());
```
Comment (GP-Dan-Tovbein) [inline assets/src/blocks/Articles/useArticlesFetch.js:63]: See [here](https://developer.mozilla.org/en-US/docs/Web/API/AbortController/abort) more information about the AbortController class

File: assets/src/blocks/Covers/useCovers.js [hunk 3]
```diff
@@ -53,26 +54,38 @@ export const useCovers = ({ post_types, tags, cover_type, initialRowsLimit, post
...
 
   useEffect(() => {
     if (!noLoading) {
-      loadCovers();
+      setController(typeof AbortController === 'undefined' ? undefined : new AbortController());
     }
   }, [cover_type, post_types, tags, posts, layout]);
 
+  useEffect(() => {
+    if(controller) {
```
Comment (GP-Dan-Tovbein) [inline assets/src/blocks/Covers/useCovers.js:76]: This ensures that the fetch request is always invoked with the `AbortController` everytime it's supported by the browser. In this case it fully supported.


 
Screenshot 2022-09-23 at 14 06 46

**Review by mleray — COMMENTED** [review]
Looks good to me as it removes the console warnings, good job 👍
However it seems that it changes the loading behaviour a bit, I've tested with a Covers block and usually switching the style to "Take Action" should load some covers without any other changes, but on this branch it doesn't happen anymore... Do you also see this happening?

**Comment by GP-Dan-Tovbein** [comment]
> Looks good to me as it removes the console warnings, good job 👍 However it seems that it changes the loading behaviour a bit, I've tested with a Covers block and usually switching the style to "Take Action" should load some covers without any other changes, but on this branch it doesn't happen anymore... Do you also see this happening?

@mleray thanks for your review!

I've recorded two different videos to show you how it's working on my local.

**This recording shows how it's working on the `master` branch.**


https://user-images.githubusercontent.com/77975803/192284516-4b4d855e-272c-418f-885a-195704fa6403.mov


**This recording shows how it's working on the current branch.**

https://user-images.githubusercontent.com/77975803/192284619-b22867c6-6f54-4ac5-817f-57aba7b41043.mov



Let me know what do you think!

**Review by mleray — APPROVED** [review]
Tested on local, looks good to me 👍

**Pull request merged**